File: #60_Permutation_Sequence/Solution.java
import java.util.*;

/* time O(n ^ 2) */

public class Solution {
    public String getPermutation(int n, int k) {
    	List<Integer> list = new LinkedList();
    	for (int i = 1; i <= n; i++) list.add(i);
        StringBuilder sb = new StringBuilder();
	    int start = 0; // inclusive
	    int kth = k - start;
	    /* O(n) */
        for (int i = 0; i < n; i++) {
        	/* O(n) */
        	int fac = factorial(n - i - 1);
        	int count = (kth - 1) / fac;
        	/* O(n) */
        	sb.append(list.remove(count));
        	start = count * fac;
        	kth = kth - start;
        }
        return sb.toString();
    }

    private int factorial(int n) {
    	int result = 1;
    	while (n > 0) {
    		result *= n;
    		n--;
    	}
    	return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.getPermutation(1, 1));
    	// System.out.println(sol.getPermutation(4, 2));
    	// System.out.println(sol.getPermutation(4, 3));
    	// System.out.println(sol.getPermutation(4, 4));
    	// System.out.println(sol.getPermutation(4, 5));
    	// System.out.println(sol.getPermutation(4, 6));
    }
}

File: #60_Permutation_Sequence/Solution.py
'''
time: O(n * n)
space: O(n)
'''

class Solution:

	def __init__(self):
		self.ans = None

	# @param {integer} n
	# @param {integer} k
	# @return {string}
	def getPermutation(self, n, k):
		if n == 0 or k == 0:
			return ''
		if n == 1:
			return '1'
		self.ans = []
		count = 1
		nums = [str(count) for count in range(1, n + 1)]
		self.KthPermutation(nums, k - 1, self.factorial(n - 1))
		return ''.join(self.ans)

	# @param {integer} nums   set of numbers
	# @param {integer} k      k index element in the array
	# @param {integer} factN  (n - 1)!
	# @return {integer[]} chosen character
	def KthPermutation(self, nums, k, factN):
		if len(nums) == 1:
			return self.ans.append(nums[0])
		nums.sort()
		self.swap(nums, 0, k / factN)
		self.ans.append(nums[0])
		self.KthPermutation(nums[1:], k - factN * (k / factN), factN / len(nums[1:]))

	def factorial(self, n):
		ans = 1
		while n > 0:
			ans *= n
			n -= 1
		return ans

	def swap(self, nums, i, j):
		if i == j:
			return
		tmp = nums[j]
		nums[j] = nums[i]
		nums[i] = tmp

sol = Solution()
print sol.getPermutation(3, 3)